### PR TITLE
Laboratory Work #1, Sergei Veselov - 20FPL3

### PR DESCRIPTION
**Pull request by Azzruge**

### CODE REVIEW

File: idk.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import requests
```
Comment (dmitry-uraev) [inline idk.py:1]: All crawler code should be written in scrapper.py.

File: scrapper.py [hunk 2]
```diff
@@ -7,6 +7,7 @@ class IncorrectURLError(Exception):
     """
     Seed URL does not match standard pattern
     """
+    #potomu chto ja pudge f
```
Comment (dmitry-uraev) [inline scrapper.py:10]: *hook*

File: scrapper.py [hunk 3]
```diff
@@ -56,8 +57,8 @@ def validate_config(crawler_path):
     Validates given config
     """
     pass
-
-
+#like if you love python
```
Comment (dmitry-uraev) [inline scrapper.py:60]: Try to avoid other unnecessary and unrelated comments in your code.

**Review by dmitry-uraev — CHANGES_REQUESTED** [review]

File: scrapper.py [hunk 4]
```diff
@@ -25,39 +33,94 @@ class Crawler:
...
+    for urls in seed_urls:
+        if urls[0:8] != 'https://' and urls[0:7] != 'http://':
+            raise IncorrectURLError
+    return seed_urls, max_articles
 
 
 if __name__ == '__main__':
-    # YOUR CODE HERE
-    pass
+    urlsi, articles = validate_config(CRAWLER_CONFIG_PATH)
```
Comment (dmitry-uraev) [inline scrapper.py:121]: good

**Review by dmitry-uraev — CHANGES_REQUESTED** [review]
Lets finish parser

File: scrapper.py [hunk 5]
```diff
@@ -25,39 +33,94 @@ class Crawler:
...
     def __init__(self, seed_urls, max_articles: int):
-        pass
+        self.seed_urls = seed_urls
+        self.max_articles = max_articles
+        self.urls = []
 
     def _extract_url(self, article_bs):
-        pass
+        class_bs = article_bs.find_all('div', class_='title')
+        for title in class_bs:
```
[the remark this answers is not in the document]
Comment (Azzruge) [reply scrapper.py:44]: so should i firstly go through the all items where url is in def find articles?

File: scrapper.py [hunk 6]
```diff
@@ -25,39 +34,117 @@ class Crawler:
...
+
+    def _fill_article_with_meta_information(self, article_bs):
+        """
+        fills article instance with title, date and topics
+        """
+        meta_bs = article_bs.find('div', class_='news_item fullnews')
+        title_bs = meta_bs.find('h1').text
+        self.article.title = title_bs
+
+        date_bs = meta_bs.find('span', class_='date').text
```
Comment (dmitry-uraev) [inline scrapper.py:87]: date can be parsed in a separate function - single responsibility principle

File: scrapper.py [hunk 7]
```diff
@@ -25,39 +34,117 @@ class Crawler:
...
+        date = datetime.strptime(date_bs, '%d.%m.%Y : %H.%M')
+        self.article.date = date
+
+        topic = article_bs.find('h4', class_='chapter_h').text
+        self.article.topics.append(topic)
+
+        self.article.author = 'NOT FOUND'
```
Comment (dmitry-uraev) [inline scrapper.py:94]: are you sure?

File: scrapper.py [hunk 8]
```diff
@@ -25,39 +34,117 @@ class Crawler:
...
+    seed_urls = config['seed_urls']
+    max_articles = config['total_articles_to_find_and_parse']
+    if not isinstance(max_articles, int) or max_articles <= 0:
+        raise IncorrectNumberOfArticlesError
+    if not isinstance(seed_urls, list) or not seed_urls:
+        raise IncorrectURLError
+    if max_articles > 100:
+        raise NumberOfArticlesOutOfRangeError
+    for urls in seed_urls:
+        if urls[0:8] != 'https://' and urls[0:7] != 'http://':
```
Comment (dmitry-uraev) [inline scrapper.py:134]: is there a more elegant way of checking this condition?

File: scrapper.py [hunk 9]
```diff
@@ -25,39 +34,117 @@ class Crawler:
...
 
 if __name__ == '__main__':
-    # YOUR CODE HERE
-    pass
+    sites, articles = validate_config(CRAWLER_CONFIG_PATH)
+    prepare_environment(ASSETS_PATH)
+    crawler = Crawler(sites, articles)
+    crawler.find_articles()
+    print(crawler.urls)
+    A_ID = 1
```
Comment (dmitry-uraev) [inline scrapper.py:145]: can you avoid such counters variables?

File: scrapper.py [hunk 10]
```diff
@@ -25,39 +34,117 @@ class Crawler:
...
+        topic = article_bs.find('h4', class_='chapter_h').text
+        self.article.topics.append(topic)
+
+        self.article.author = 'NOT FOUND'
+
+    def parse(self):
+        """
+        does all the work
+        """
+        response = requests.get(self.article_url)
```
Comment (dmitry-uraev) [inline scrapper.py:100]: no headers or other settings. Can you collect with such setup 100 articles safely?

**Review by dmitry-uraev — CHANGES_REQUESTED** [review]
Good job. Need some fixes and we are done here

File: scrapper.py (hunk 6 above)
Comment (Azzruge) [reply scrapper.py:87]: Why can't it collect all the metadata in one function? I thought it all one responsibility

File: scrapper.py (hunk 7 above)
Comment (Azzruge) [reply scrapper.py:94]: I'm sure that there is no author in these artcles

File: scrapper.py (hunk 10 above)
Comment (Azzruge) [reply scrapper.py:100]: yes, elista.org has no DDOS defense

**Review by dmitry-uraev — COMMENTED** [review]
Well done. The second lab is ready for assignment.

**Pull request closed without merging**